allneeds.lua

```diff
@@ -0,0 +1,81 @@
+-- print sum of all needs
+--[====[
+
+allneeds
+========
+Show which needs are high for all dwarfs.
+Weight and Focus are divided by the number of dwarves that have that need.
+
+]====]
+local ENUM = {}
+
+ENUM[0]  = "Socialize"
+ENUM[1]  = "DrinkAlcohol"
+ENUM[2]  = "PrayOrMedidate"
+ENUM[3]  = "StayOccupied"
+ENUM[4]  = "BeCreative"
+ENUM[5]  = "Excitement"
+ENUM[6]  = "LearnSomething"
+ENUM[7]  = "BeWithFamily"
+ENUM[8]  = "BeWithFriends"
+ENUM[9]  = "HearEloquence"
+ENUM[10] = "UpholdTradition"
+ENUM[11] = "SelfExamination"
+ENUM[12] = "MakeMerry"
+ENUM[13] = "CraftObject"
+ENUM[14] = "MartialTraining"
+ENUM[15] = "PracticeSkill"
+ENUM[16] = "TakeItEasy"
+ENUM[17] = "MakeRomance"
+ENUM[18] = "SeeAnimal"
+ENUM[19] = "SeeGreatBeast"
+ENUM[20] = "AcquireObject"
+ENUM[21] = "EatGoodMeal"
+ENUM[22] = "Fight"
+ENUM[23] = "CauseTrouble"
+ENUM[24] = "Argue"
+ENUM[25] = "BeExtravagant"
+ENUM[26] = "Wander"
+ENUM[27] = "HelpSomebody"
+ENUM[28] = "ThinkAbstractly"
+ENUM[29] = "AdmireArt"
+
+
+local need = {}
+local n = 0
+for _, unit in ipairs(df.global.world.units.all) do
+    if not unit then
+      qerror('Unit not real.')
+    end
+    -- Select valid dwarf
+    if unit.status.current_soul and unit.race == 572 then
```

Comment:
I'm guessing 572 is the creature raw ID for dwarves in the specific world you were testing? This is **not** guaranteed to be consistent in any other save. It might be usually consistent if you're using the same raws, but any mods will break this check entirely.

There are many functions in the [Units module](https://dfhack.readthedocs.io/en/stable/docs/Lua%20API.html#units-module) that you can use to check traits of units. I would guess `isCitizen` is probably what you actually want here, although perhaps `isDwarf` if you want to exclude non-dwarven citizens from this check.